app/helpers/Emails/Notifications/Submissions/SubmissionEmailsSender.php

```diff
@@ -69,31 +70,45 @@ public function submissionEvaluated(AssignmentSolutionSubmission $submission): b
       return false;
     }
 
-    // Handle evaluation notification...
+    // Handle evaluation notifications...
+    return $this->handleEvaluationNotifications($solution, $submission, $user) &&
+      $this->handleTeacherNotifications($solution, $submission, $user);
+  }
+
+  /**
+   * @throws InvalidStateException
+   */
+  private function handleEvaluationNotifications(AssignmentSolution $solution, AssignmentSolutionSubmission $submission, User $user): bool {
     $threshold = (new DateTime())->modify($this->submissionNotificationThreshold);
     if ($user->getSettings()->getSubmissionEvaluatedEmails() && $solution->getSolution()->getCreatedAt() < $threshold) {
       $locale = $user->getSettings()->getDefaultLanguage();
-      $result = $this->createSubmissionEvaluated($assignment, $submission, $locale);
-  
+      $result = $this->createSubmissionEvaluated($solution->getAssignment(), $submission, $locale);
+
       // Send the mail
       return $this->emailHelper->send(
         $this->sender,
         [$user->getEmail()],
         $locale,
         $result->getSubject(),
         $result->getText()
-      );  
+      );
     }
 
-    // Handle teacher notifications...
-    $best = $this->assignmentSolutions->findBestSolution($assignment, $user);
-    $recipients = $this->getGroupSupervisorsRecipients($assignment->getGroup());
+    return true;
+  }
+
+  private function handleTeacherNotifications(AssignmentSolution $solution, AssignmentSolutionSubmission $submission, User $user): bool {
+    $success = true;
```

Comment:
I think this $success thing is an obfuscation which made it more difficult for me to understand the code. Furthermore, if we want to short-circuit, we should use return instead since the code is in a function. If not, this should be written in a different manner, so that reviewed email is sent even if accepted mail fails.